design/0004-multicluster-phase-2.md

```diff
@@ -0,0 +1,262 @@
+  clusters.
+* Code simplification: a single Go process no longer needs to maintain nested
+  watches where a watch on secrets starts and stops watches on target clusters.
+  Each process is responsible for a single target cluster, greatly simplifying
+  the code.
+
+The downsides of this approach are that:
+* Running more service mirror pods incurs more resource overhead, but this
+  should be negligable.
+* Running more service mirror deployments incurs more operational overhead
+  since these deployments need to be monitored and maintained.  Thankfully,
+  Prometheus makes monitoring a collection of processes just as easy as
+  monitoring a single process, assuming we use labels in a consistent way.  Any
+  observability tooling or dashboards that we build for service mirror
+  observability should have no problem dealing with metrics from multiple
+  processes.  We can also develop tooling to make upgrading easier, as discussed
+  later.
+
+## UX
+
+This proposal makes very few changes to the UX flow.
+
+
+Install the multicluster gateway, service account, and `link` CRD in both clusters.
+
+```
+linkerd --context=target multicluster install | kubectl --context=target apply -f -
+linkerd --context=source multicluster install | kubectl --context=source apply -f -
+```
+
+Link the clusters.  This will install the cluster credentails secret, `link`
+reousrce, and service mirror controller in the source cluster.
+
+```
+linkerd --context=target multicluster link --cluster-name=foobar | kubectl --context=source apply -f -
+```
+
+This will create:
+* secret/cluster-credentials-foobar
+* link.multicluster.linkerd.io/foobar
+* deployment/service-mirror-foobar
+* (also the necessary RBAC for the service mirror controller)
+
+The generated `link` resource will specify the default label selector of
+`multicluster.linkerd.io/exported=true`.
+Alternatively, users can specify their own label selector:
+
+```
+linkerd --context=target multicluster link --cluster-name=foobar -l "export-me-please=true" | kubectl --context=source apply -f -
+```
+
+Now, to actually export the service, it's just a matter of applying the correct label:
+
+```
+kubectl --context=target label svc/my-cool-svc multicluster.linkerd.io/exported=true
+```
+
+The `service-mirror-foobar` service mirror controller in the source cluster will
+see that a service in the remote cluster matches its label selector and will
+mirror that service by creating a `my-cool-svc-foobar` mirror service in the
+source cluster.
+
+Active links can be seen using `kubectl get`.  This replaces the `multicluster gateways` command:
+
+
+```
+kubectl --context=source get links
+NAME   NUM SVC   ALIVE   LATENCY_P50   LATENCY_P95   LATENCY_P99
+foobar       3    True         275ms         385ms         397ms
+```
+
+
+## Upgrading
+
+To upgrade the service mirror controller, simply run the link step again with a
+newer version of the CLI:
```

Comment:
What are the dependencies _between_ clusters? I.e. if i've got a gateway (target) cluster at version N and a mirror (source) cluster at version M, what are the things that must be common to both versions? Could these change? How will they change?

I also think this is oversimplified. How do you upgrade the CRD, for instance? Can't we "break" existing mirror configurations with changes to this CRD?

Reply:
It's hard to speculate about dependencies between clusters without knowing in advance of WHAT will change.  For example, the `link` command extracts certain information from the target cluster such as the gateway address and the service account token.  If the location or format of this information has changed, this will not work.  

I suspect that you would have to upgrade the target cluster first in this case by running `linkerd mc install` before running `link` again.  But I'm not sure this speculation is very useful without knowing in advance what is changing.

We avoid using generated code to interact with the CRD specifically to avoid issues during upgrades.  The service mirror controller will read the `link` resource using a dynamic client that gets the contents as an untyped map.  This allows us to build in our own backwards compatibility and ensure that if the format of the CRD changes that the service mirror controller is able to handle both the old and new format.